File: drive-java/src/main/java/frc/robot/commands/drive/DriveMeasure.java
package frc.robot.commands.drive;

import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.Robot;
import frc.robot.subsystems.DriveSubsystem;
import frc.robot.subsystems.DriveSubsystem.Tread;

/**
 * Utility command that measures drive distance and rotation (zeros each time it
 * is started).
 */
public class DriveMeasure extends Command {

  private DriveSubsystem drive;
  private Tread left;
  private Tread right;
  private double yawLast;
  private double leftDistLast;
  private int leftCntsLast;
  private double rightDistLast;
  private int rightCntsLast;

  public DriveMeasure() {
    super("Measure");
    // IMPORTANT. Since we are only reading data and not manipulting mechanisms, we
    // do not need
    // to require the drive subsystem
    drive = Robot.drive;
    left = drive.getLeft();
    right = drive.getRight();
    setRunWhenDisabled(true);
  }

  /**
   * Takes initial values of things measurements to "zero" each time the command
   * is started.
   * 
   */
  @Override
  protected void initialize() {
    yawLast = drive.getAngle();
    leftDistLast = left.getDistance();
    leftCntsLast = left.getCounts();
    rightDistLast = left.getDistance();
    rightCntsLast = right.getCounts();
  }

  // Called repeatedly when this Command is scheduled to run
  @Override
  protected void execute() {
    double yawRaw = drive.getAngle();
    double yaw = yawRaw - yawLast;
    double leftDist = left.getDistance() - leftDistLast;
    int leftCnts = left.getCounts() - leftCntsLast;
    double leftVel = left.getVelocity();
    double rightDist = right.getDistance() - rightDistLast;
    int rightCnts = right.getCounts() - rightCntsLast;
    double rightVel = right.getVelocity();

    SmartDashboard.putNumber("Yaw NavX", yawRaw);
    SmartDashboard.putNumber("Yaw Measured", yaw);
    SmartDashboard.putNumber("Left Dist", leftDist);
    SmartDashboard.putNumber("Left Cnts", leftCnts);
    SmartDashboard.putNumber("Left Vel", leftVel);
    SmartDashboard.putNumber("Right Dist", rightDist);
    SmartDashboard.putNumber("Right Cnts", rightCnts);
    SmartDashboard.putNumber("Right Vel", rightVel);

    SmartDashboard.putNumber("Accel X", Robot.drive.getAccelX());
    SmartDashboard.putNumber("Accel Y", Robot.drive.getAccelY());
    SmartDashboard.putBoolean("Bump", Robot.drive.bumpCheck(0.4, 0.4));
  }

  @Override
  protected boolean isFinished() {
    return false;
  }
}

File: drive-cpp/src/main/cpp/commands/drive/DriveMeasure.cpp
#include "commands/drive/DriveMeasure.h"
#include "Robot.h"

#include <frc/smartdashboard/SmartDashboard.h>

using namespace frc;

DriveMeasure::DriveMeasure() : Command("Measure") {
  // IMPORTANT. Since we are only reading data and not manipulting mechanisms, we
  // do not need
  // to require the drive subsystem
  SetRunWhenDisabled(true);
}

void DriveMeasure::Initialize() {
  DriveSubsystem& drive = Robot::drive;
  DriveSubsystem::Tread& left = drive.getLeft();
  DriveSubsystem::Tread& right = drive.getRight();

  yawLast = drive.getAngle();
  leftDistLast = left.getDistance();
  leftCntsLast = left.getCounts();
  rightDistLast = left.getDistance();
  rightCntsLast = right.getCounts();
}

void DriveMeasure::Execute() {
  DriveSubsystem& drive = Robot::drive;
  DriveSubsystem::Tread& left = drive.getLeft();
  DriveSubsystem::Tread& right = drive.getRight();

  double yawRaw = drive.getAngle();
  double yaw = yawRaw - yawLast;
  double leftDist = left.getDistance() - leftDistLast;
  int leftCnts = left.getCounts() - leftCntsLast;
  double leftVel = left.getVelocity();
  double rightDist = right.getDistance() - rightDistLast;
  int rightCnts = right.getCounts() - rightCntsLast;
  double rightVel = right.getVelocity();

  SmartDashboard::PutNumber("Yaw NavX", yawRaw);
  SmartDashboard::PutNumber("Yaw Measured", yaw);
  SmartDashboard::PutNumber("Left Dist", leftDist);
  SmartDashboard::PutNumber("Left Cnts", leftCnts);
  SmartDashboard::PutNumber("Left Vel", leftVel);
  SmartDashboard::PutNumber("Right Dist", rightDist);
  SmartDashboard::PutNumber("Right Cnts", rightCnts);
  SmartDashboard::PutNumber("Right Vel", rightVel);

  SmartDashboard::PutNumber("Accel X", drive.getAccelX());
  SmartDashboard::PutNumber("Accel Y", drive.getAccelY());
  SmartDashboard::PutBoolean("Bump", drive.bumpCheck(0.4, 0.4));
}

bool DriveMeasure::IsFinished() {
  return false;
}
